### PR TITLE
APIv4 - Add Contact primary/billing joins for email, address, phone & IM

### PR DESCRIPTION
**Pull request by colemanw**
Overview
----------------------------------------
Replaces #23819 based on feedback.

Declares an implicit join between the contact record and primary/billing email, phone, address & im records,
making it easier to retrieve those directly from the Contact API.

### CODE REVIEW

File: Civi/Api4/Event/Subscriber/ContactSchemaMapSubscriber.php [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+  }
+
+  /**
+   * @param \Civi\Api4\Event\SchemaMapBuildEvent $event
+   */
+  public function onSchemaBuild(SchemaMapBuildEvent $event) {
+    $schema = $event->getSchemaMap();
+    $table = $schema->getTableByName('civicrm_contact');
+
+    // Add links to primary email, address, phone & im
```
Comment (eileenmcnaughton) [inline Civi/Api4/Event/Subscriber/ContactSchemaMapSubscriber.php:28]: There is ALSO `openid` - that is the bane of my life. I would be happy to see us not cater to that OR `im` which I really doubt people use - but I could be wrong - but completeness would include both

**Comment by eileenmcnaughton** [comment]
style -

image


Also - I REALLY want to do billing and primary

File: Civi/Api4/Event/Subscriber/ContactSchemaMapSubscriber.php (hunk 1 above)
Comment (colemanw) [reply Civi/Api4/Event/Subscriber/ContactSchemaMapSubscriber.php:28]: People do use IM, although that term has a bit of a retro feel, there are still plenty of uses for it. civicrm.org uses it to record a person's mattermost username.

**Comment by colemanw** [comment]
@eileenmcnaughton tests are passing and I'm not inclined to expand the scope within this PR but we could follow up with more after this is merged.

**Comment by eileenmcnaughton** [comment]
api\v4\Action\FkJoinTest::testOptionalJoin
PEAR_Exception: DB Error: syntax error

/home/jenkins/bknix-dfl/build/core-23845-4e8ci/web/sites/all/modules/civicrm/CRM/Core/Error.php:955
/home/jenkins/bknix-dfl/build/core-23845-4e8ci/web/sites/all/modules/civicrm/vendor/pear/pear-core-minimal/src/PEAR.php:922
/home/jenkins/bknix-dfl/build/core-23845-4e8ci/web/sites/all/modules/civicrm/vendor/pear/db/DB.php:997
/home/jenkins/bknix-dfl/build/core-23845-4e8ci/web/sites/all/modules/civicrm/vendor/pear/pear-core-minimal/src/PEAR.php:575
/home/jenkins/bknix-dfl/build/core-23845-4e8ci/web/sites/all/modules/civicrm/vendor/pear/pear-core-minimal/src/PEAR.php:223
/home/jenkins/bknix-dfl/build/core-23845-4e8ci/web/sites/all/modules/civicrm/vendor/pear/db/DB/common.php:1928
/home/jenkins/bknix-dfl/build/core-23845-4e8ci/web/sites/all/modules/civicrm/vendor/pear/db/DB/mysqli.php:936
/home/jenkins/bknix-dfl/build/core-23845-4e8ci/web/sites/all/modules/civicrm/vendor/pear/db/DB/mysqli.php:406
/home/jenkins/bknix-dfl/build/core-23845-4e8ci/web/sites/all/modules/civicrm/vendor/pear/db/DB/common.php:1234
/home/jenkins/bknix-dfl/build/core-23845-4e8ci/web/sites/all/modules/civicrm/packages/DB/DataObject.php:2696
/home/jenkins/bknix-dfl/build/core-23845-4e8ci/web/sites/all/modules/civicrm/packages/DB/DataObject.php:1829
/home/jenkins/bknix-dfl/build/core-23845-4e8ci/web/sites/all/modules/civicrm/CRM/Core/DAO.php:472
/home/jenkins/bknix-dfl/build/core-23845-4e8ci/web/sites/all/modules/civicrm/CRM/Core/DAO.php:1637
/home/jenkins/bknix-dfl/build/core-23845-4e8ci/web/sites/all/modules/civicrm/Civi/Api4/Query/Api4SelectQuery.php:167
/home/jenkins/bknix-dfl/build/core-23845-4e8ci/web/sites/all/modules/civicrm/Civi/Api4/Generic/DAOGetAction.php:114
/home/jenkins/bknix-dfl/build/core-23845-4e8ci/web/sites/all/modules/civicrm/Civi/Api4/Generic/DAOGetAction.php:101
/home/jenkins/bknix-dfl/build/core-23845-4e8ci/web/sites/all/modules/civicrm/Civi/Api4/Provider/ActionObjectProvider.php:69
/home/jenkins/bknix-dfl/build/core-23845-4e8ci/web/sites/all/modules/civicrm/Civi/API/Kernel.php:149
/home/jenkins/bknix-dfl/build/core-23845-4e8ci/web/sites/all/modules/civicrm/Civi/Api4/Generic/AbstractAction.php:234
/home/jenkins/bknix-dfl/build/core-23845-4e8ci/web/sites/all/modules/civicrm/tests/phpunit/api/v4/Action/FkJoinTest.php:88
/home/jenkins/bknix-dfl/extern/phpunit8/phpunit8.phar:671

**Comment by eileenmcnaughton** [comment]
I think this might be easy to fix up @colemanw

**Comment by colemanw** [comment]
@eileenmcnaughton I've updated this to create virtual fields for both primary and billing.

I doubt we'll ever do this for the location types. Too many issues with that, foremost is that there is no unique constraint on them in the db so you can create e.g. multiple "home" email addresses.

**Comment by eileenmcnaughton** [comment]
Not sure what this is about

api\v4\Custom\CustomGroupACLTest::testViewEditCustomGroupACLs
Trying to access array offset on value of type bool

/home/jenkins/bknix-dfl/build/core-23845-4ufof/web/sites/all/modules/civicrm/Civi/Api4/Query/Api4SelectQuery.php:802
/home/jenkins/bknix-dfl/build/core-23845-4ufof/web/sites/all/modules/civicrm/Civi/Api4/Query/Api4SelectQuery.php:762
/home/jenkins/bknix-dfl/build/core-23845-4ufof/web/sites/all/modules/civicrm/Civi/Api4/Query/Api4SelectQuery.php:125
/home/jenkins/bknix-dfl/build/core-23845-4ufof/web/sites/all/modules/civicrm/Civi/Api4/Generic/DAOGetAction.php:113
/home/jenkins/bknix-dfl/build/core-23845-4ufof/web/sites/all/modules/civicrm/Civi/Api4/Generic/DAOGetAction.php:101
/home/jenkins/bknix-dfl/build/core-23845-4ufof/web/sites/all/modules/civicrm/Civi/Api4/Provider/ActionObjectProvider.php:69
/home/jenkins/bknix-dfl/build/core-23845-4ufof/web/sites/all/modules/civicrm/Civi/API/Kernel.php:149
/home/jenkins/bknix-dfl/build/core-23845-4ufof/web/sites/all/modules/civicrm/Civi/Api4/Generic/AbstractAction.php:234
/home/jenkins/bknix-dfl/build/core-23845-4ufof/web/sites/all/modules/civicrm/tests/phpunit/api/v4/Custom/CustomGroupACLTest.php:196
/home/jenkins/bknix-dfl/extern/phpunit8/phpunit8.phar:671

test this please

**Comment by eileenmcnaughton** [comment]
Hmm I got an error

image


image



Still digging as  I think it might be the explorer not the api

**Comment by eileenmcnaughton** [comment]
This doesn't error

image

But since I got all empty results I tried

image


which does error
image

**Comment by eileenmcnaughton** [comment]
Hmm - no result on a contact which definitely SHOULD have one

image

**Comment by eileenmcnaughton** [comment]
Also note that we have introduced tokens for these fields for `primary_address` - which are handed in `CRM_Contact_Tokens` "as if" the functionality existed in apiv4 - if we want to switch the order from  `primary_address`  to ` address_primary` then we should do that now - but note it would be a lot let work to use  `primary_address`  & not have to upgrade any existing tokens

image

**Comment by colemanw** [comment]
@eileenmcnaughton I saw the same problem until I did `civibuild rebuild`. I think there's some deep caching going on with these containers, but I promise it really does work, as the unit test demonstrates :)

**Comment by eileenmcnaughton** [comment]
@colemanw it worked this time - first use of the day so maybe caches were timed out - but what is the intent behind 'primary_email'  (as a field in itself)

image


image

**Comment by colemanw** [comment]
@eileenmcnaughton That is the FK which allows the join to work.

**Comment by eileenmcnaughton** [comment]
@colemanw OK - so the FK shows up as a selectable field that aliases to id as well as the fields? Are there other places we do that?

**Comment by colemanw** [comment]
> Are there other places we do that?

Yes, that's the way all FK fields work in the API. For example you can do:
image

And it will return the same value for both.

**Pull request closed without merging**